### PR TITLE
Automatic shape inference !

### PR DESCRIPTION
**Pull request by farizrahman4u**
Infer output shape for any theano / TF expression or function.

 **Examples :**

* Infering shape for K.concatenate : 
```python

from keras import backend as K

shape1 = (None, 20, 10)
shape2 = (None, 20, 30)

output_shape = K.infer_shape(K.concatenate, [shape1, shape2])
print output_shape  # >>> (None, 20, 40)
```

* Lambda layer + merge with shape inference : 

```python


def f(x):
    # sum along time axis
    return K.sum(x, axis=1)

def m(X):
    x = X[0]
    y = X[1]
    return 2*x + 3*y

a = Input((20, 10))
b = Input(20, 10))

c = merge([a, b], mode=m)  # no output shape required
d = Lambda(f)(c)
print d._keras_shape  #  >>> (None, 10)
```

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -611,6 +611,32 @@ def gradients(loss, variables):
...
+        input_shape = [input_shape]
+    xs = [placeholder(shape=shape) for shape in input_shape]
+    outputs = func(xs if len(xs) > 1 else xs[0])
+    if type(outputs) is not list:
+        outputs = [outputs]
+    output_shapes = [int_shape(o) for o in outputs]
+    if len(output_shapes) == 1:
+        return output_shapes[0]
+    else:
+        return output_shapes
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I feel like it would actually be clearer and more economical to separate the two cases entirely: one input vs. multiple inputs.

**Comment by fchollet** [comment]
Cool, this looks useful!

File: keras/backend/theano_backend.py [hunk 2]
```diff
@@ -501,6 +501,50 @@ def function(inputs, outputs, updates=[], **kwargs):
...
+        ys = [ys]
+    f = function(inputs=xs, outputs=[y.shape for y in ys])
+    output_shapes_1 = list(map(list, f(input1)))
+    output_shapes_2 = list(map(list, f(input2)))
+    for i in range(len(output_shapes_1)):
+        for j in range(len(output_shapes_1[i])):
+            if output_shapes_1[i][j] != output_shapes_2[i][j]:
+                output_shapes_1[i][j] = None
+    output_shapes_1 = list(map(tuple, output_shapes_1))
+    if len(output_shapes_1) == 1:
```
Comment (farizrahman4u) [inline keras/backend/theano_backend.py]: This is a mess.

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -501,6 +501,49 @@ def function(inputs, outputs, updates=[], **kwargs):
...
+    xs = [placeholder(ndim=len(input_shapes[i])) for i in range(len(input_shapes))]
+    ys = func(xs if len(xs) > 1 else xs[0])
+    if type(ys) not in [list, tuple]:
+        ys = [ys]
+    f = function(inputs=xs, outputs=[y.shape for y in ys])
+    output_shapes_1 = list(map(list, f(input1)))
+    output_shapes_2 = list(map(list, f(input2)))
+    for i in range(len(output_shapes_1)):
+        for j in range(len(output_shapes_1[i])):
+            if output_shapes_1[i][j] != output_shapes_2[i][j]:
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: I fear this is a brittle mechanism. It will work in simple cases but will fail in advanced cases.

File: keras/backend/theano_backend.py [hunk 4]
```diff
@@ -501,6 +501,49 @@ def function(inputs, outputs, updates=[], **kwargs):
...
+        input_shapes = [input_shape]
+    else:
+        input_shapes = input_shape
+    input_shapes_1 = list(map(list, input_shapes))
+    input_shapes_2 = list(map(list, input_shapes))
+    for i in range(len(input_shapes)):
+        for j in range(len(input_shapes[i])):
+            if not input_shapes_1[i][j]:
+                input_shapes_1[i][j] = 10
+                input_shapes_2[i][j] = 0
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: I don't think that `0` is a valid shape dimension.

**Comment by fchollet** [comment]
We need to check with the Theano devs to what extent it's possible to do shape inference outside of the graph in Theano.

I think relying on filler values for `None` dimensions is not a reliable mechanism.

File: keras/backend/theano_backend.py (hunk 4 above)
Comment (farizrahman4u) [inline keras/backend/theano_backend.py]: If we set it to, say 20, then x[:5] would give the same shape for both input shapes, and will not be set to None.

**Comment by farizrahman4u** [comment]
Can you provide an edge case where this breaks?

**Comment by farizrahman4u** [comment]
Consider `input_shape` (X, 1, 2, 3, ..), where `X` is an unknown dimension. The code can only be broken by a function which gives the same `output_shape` for `X=10` and `X=0` but not for all values of X. Such a function is impossible right?

**Comment by fchollet** [comment]
> The code can only be broken by a function which gives the same output_shape for X=10 and X=0 but not for all values of X. Such a function is impossible right?

It would trivial to build a pathological case.
```python
def break_your_code(x):
   return T.switch(x.shape[0] % 10, x, T.concatenate([x, x]))
```

**Comment by farizrahman4u** [comment]
Cool! didn't think about functions that use shape in the calculation.

**Comment by farizrahman4u** [comment]
Inferring shapes of such functions seems to be pretty hard (otherwise theano would already have some built in function to do it). In the context of deep learning, I am not aware of any case where the function depends on the shape tensor of the input (the gradients won't make sense). Maybe we could add a warning in the doc string regarding this. In such edge cases, the user can override the `output_shape` of `Lambda` and `merge`. I have some upcoming PRs that require shape inference. What say?

**Comment by fchollet** [comment]
I think we should consult with the Theano devs first, they may have some interesting fresh ideas about shape inference outside of the graph.

Otherwise, we can go with a partial solution as long as it's properly documented, e.g. only handling complete input shapes (except for the batch dimension). Not a fan of the 2 filler values trick.

**Comment by farizrahman4u** [comment]
It wouldn't make a difference, Even if there is only one `None` in the input shape, in the output_shape it could cause multiple `None`s, or the `None` could just move around with `K.permute_dimensions`.Also people use Keras as a general tensor library in non - deep learning contexts where there is no notion of a batch_size, so `K` functions are expected to handle arbitrary tensors, not just NN inputs with well defined batch_size, timesteps etc

**Comment by farizrahman4u** [comment]
@nouiz help?

**Comment by ErmiaAzarkhalili** [comment]
I am curious to know why you didn't implement your idea into keras? it was great... I read the comments, but cannot understand what the main issue was? Theano?
